hal/src/nRF52840/ble_hal.cpp

```diff
@@ -15,739 +15,3147 @@
+    hal_ble_addr_t connectingAddr_;                 /**< Address of peer the Central is connecting to. */
+    uint16_t disconnectingHandle_;                  /**< Handle of connection to be disconnected. */
+    uint16_t connParamUpdateHandle_;                /**< Handle of the connection that is to send peripheral connection update request. */
+    uint8_t connParamsUpdateAttempts_;              /**< Attempts for peripheral to update connection parameters. */
+    os_timer_t connParamsUpdateTimer_;              /**< Timer used for sending peripheral connection update request after connection established. */
+    os_semaphore_t connParamsUpdateSemaphore_;      /**< Semaphore to wait until connection parameters updated. */
+    os_semaphore_t connectSemaphore_;               /**< Semaphore to wait until connection established. */
+    os_semaphore_t disconnectSemaphore_;            /**< Semaphore to wait until connection disconnected. */
+    Vector<BleConnection_t> connections_;           /**< Current on-going connections. */
+    BleObject* bleObj_;                             /**< Pointer to the BLE Object. */
+
+    static const uint32_t CONNECTION_OPERATION_TIMEOUT_MS = 5000;
+
+    BleConnection_t* fetchConnection(uint16_t connHandle);
+    BleConnection_t* fetchConnection(const hal_ble_addr_t* address);
+    void addConnection(const BleConnection_t& connection);
+    void removeConnection(uint16_t connHandle);
+    void initiateConnParamsUpdateIfNeeded(const BleConnection_t* connection);
+    bool isConnParamsFeeded(const hal_ble_conn_params_t* params) const;
+    static void toPlatformConnParams(ble_gap_conn_params_t* params, const hal_ble_conn_params_t* halConnParams);
+    static void toHalConnParams(hal_ble_conn_params_t* halConnParams, const ble_gap_conn_params_t* params);
+    static void onConnParamsUpdateTimerExpired(os_timer_t timer);
+    static void processConnectionEvents(const ble_evt_t* event, void* context);
 };
 
-// Current profile
-Profile g_profile = {};
+class BleObject::GattServer : public GattBase {
+public:
+    GattServer(BleObject* obj);
+    ~GattServer() = default;
+    int init(void);
+    int addService(uint8_t type, const hal_ble_uuid_t* uuid, uint16_t* handle);
+    int addCharacteristic(uint16_t svcHandle, uint8_t properties, const hal_ble_uuid_t* uuid, const char* desc, hal_ble_char_handles_t* handles);
+    int addDescriptor(uint16_t charHandle, const hal_ble_uuid_t* uuid, uint8_t* descriptor, uint16_t len, uint16_t* handle);
+    size_t setValue(uint16_t handle, const uint8_t* buf, size_t len);
+    size_t getValue(uint16_t handle, uint8_t* buf, size_t len);
+
+private:
+    class BleCharacteristic {
+        public:
+            uint8_t properties;
+            uint16_t svcHandle;
+            hal_ble_char_handles_t handles;
+            Vector<uint16_t> cccdConnections;
+    };
+
+    static const uint32_t BLE_HVX_PROCEDURE_TIMEOUT_MS = 5000;
```

Comment:
Consider moving this out of class into the anonymous namespace in the top of the source file for better visibility.